router/sources/router.move

```diff
@@ -0,0 +1,928 @@
+    /// @param target_addr The address the registered name will point to
+    /// @param to_addr The address to send the token to. If none, then the user will be the owner. In MODE_V1, receiver must have already opted in to direct_transfer via token::opt_in_direct_transfer
+    public entry fun register_domain(
+        user: &signer,
+        domain_name: String,
+        registration_duration_secs: u64,
+        target_addr: Option<address>,
+        to_addr: Option<address>,
+    ) acquires RouterConfig {
+        let mode = get_mode();
+        if (mode == MODE_V1) {
+            assert!(
+                registration_duration_secs % SECONDS_PER_YEAR == 0,
+                error::invalid_argument(ENOT_MULTIPLE_OF_SECONDS_PER_YEAR)
+            );
+            domains::register_domain(
+                user,
+                domain_name,
+                ((registration_duration_secs / SECONDS_PER_YEAR) as u8)
+            );
+        } else if (mode == MODE_V1_AND_V2) {
+            assert!(can_register_in_v2(domain_name, option::none()), error::unavailable(ENAME_NOT_AVAILABLE));
+            v2_domains::register_domain(
+                &get_router_signer(),
+                user,
+                domain_name,
+                registration_duration_secs,
+            );
+            // Clear the name in v1
+            domains::force_clear_registration(&get_router_signer(), option::none(), domain_name)
+        } else {
+            abort error::not_implemented(ENOT_IMPLEMENTED_IN_MODE)
+        };
+
+        // Common operations that handle modes via the router
+        if (option::is_some(&target_addr)) {
+            set_target_addr(
+                user,
+                domain_name,
+                option::none(),
+                *option::borrow(&target_addr)
+            );
+        };
+        if (option::is_some(&to_addr)) {
+            transfer_name(user, domain_name, option::none(), *option::borrow(&to_addr));
+        };
+
+        set_primary_name_when_register(
+            user,
+            target_addr,
+            to_addr,
+            domain_name,
+            option::none(),
+        );
+    }
+
+    fun set_primary_name_when_register(
+        user: &signer,
+        target_addr: Option<address>,
+        to_addr: Option<address>,
+        domain_name: String,
+        subdomain_name: Option<String>,
+    ) acquires RouterConfig {
+        let owner_addr = address_of(user);
+
+        // if the owner address is not the buyer address
+        if (option::is_some(&to_addr) && to_addr != option::some(owner_addr)) {
+            return
+        };
+
+        // if the target address is not the buyer address
+        if (option::is_some(&target_addr) && target_addr != option::some(owner_addr)) {
+            return
+        };
+
+        let has_primary_name = has_primary_name(user);
+        if (!has_primary_name) {
+            set_primary_name(user, domain_name, subdomain_name);
```

Comment:
Why does set_primary_name also migrate the domain as well instead of just setting the primary name? This also means register_domain also does the migration at the same time, which is overlapping with migrate_name. Why not just disallow that and only allow migrate_name to do the migration to avoid extra complexity and branches? One function one task = much easier to reason, test and debug

Reply:
This does create a lot of edge cases in all of the interactions though. It'd be strange for a user, for example, to want to change the primary name and not expecting anything but that. But to their surprise, the domain NFT disappears from their wallet (and a new one randomly appears). Is there a good solution for this on the FE side?